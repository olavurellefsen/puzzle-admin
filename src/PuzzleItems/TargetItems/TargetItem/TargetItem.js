import React, { useState, useRef } from "react";
import {
  TargetItemContainer,
  TargetItemForm,
  TargetItemFieldCaption,
  TargetItemInputField
} from "./TargetItem.style";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";

export default props => {
  const { targetitem } = props;
  const puzzleItem = targetitem.puzzleItemBypuzzleItemId;
  const puzzleitemId = puzzleItem.id;
  const puzzleitemName = puzzleItem.name;
  const puzzleItemText = puzzleItem.itemBypuzzletextKey.value;
  const puzzletextKey = puzzleItem.puzzletext_key;

  const [textField, setTextField] = useState(puzzleItemText);
  const [puzzleItemNameField, setPuzzleItemNameField] = useState(puzzleitemName);
  
  const itemText = useRef(null);
  const itemName = useRef(null);

  // If the parent component is updated (someone else has changed the value), then update the field here
  if (puzzleItemText!==textField && puzzleitemName!==puzzleItemNameField) {
    if (itemText.current.defaultValue === textField) {
      setTextField(puzzleItemText);
    }
    if (itemName.current.defaultValue === puzzleItemNameField) {
      setPuzzleItemNameField(puzzleitemName);
    }    
  }

  const UPDATE_DRAGITEM = gql`
    mutation update_targetitem(
      $puzzleitemid: Int!
      $puzzleitemtext: String!
      $puzzletextitemkey: String!
      $puzzlename: String!
    ) {
      update_puzzle_item(where: { id: { _eq: $puzzleitemid } }, _set: { name: $puzzlename }) {
        affected_rows
      }
      update_item(
        where: {
          _and: [
            { key: { _eq: $puzzletextitemkey } }
            { language_id: { _eq: 1 } }
          ]
        }
        _set: { value: $puzzleitemtext }
      ) {
        affected_rows
      }
    }
  `;

  return (
    <TargetItemContainer
      data-testid="TargetItemContainer"
      className={`target${targetitem.id}`}
      id={`target${targetitem.id}`}
    >
      <Mutation mutation={UPDATE_DRAGITEM}>
        {(updateTargetItem, { data }) => {
          const submitForm = e => {
            e.preventDefault();
            updateTargetItem({
              variables: {
                puzzleitemid: puzzleitemId,
                puzzleitemtext: textField,
                puzzletextitemkey: puzzletextKey,
                puzzlename: puzzleItemNameField
              }
            });
            itemName.current.blur();
            itemText.current.blur();
          };
          return (
            <TargetItemForm
              onKeyPress={e => {
                if (e.key === "Enter") {
                  submitForm(e);
                }
              }}
              onSubmit={e => submitForm(e)}
            >
              <TargetItemFieldCaption>LÝSING AV HÁLIMYND {targetitem.id}</TargetItemFieldCaption>
              <TargetItemInputField
                ref={itemName}
                name="itemName"
                type="text"
                value={puzzleItemNameField}
                onChange={e => setPuzzleItemNameField(e.target.value)}
              />
              <TargetItemFieldCaption>TEKSTUR</TargetItemFieldCaption>
              <TargetItemInputField
                ref={itemText}
                name="itemText"
                type="text"
                value={textField}
                onChange={e => setTextField(e.target.value)}
              />
            </TargetItemForm>
          );
        }}
      </Mutation>
    </TargetItemContainer>
  );
};
