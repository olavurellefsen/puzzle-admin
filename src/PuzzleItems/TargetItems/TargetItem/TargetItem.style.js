import styled from 'styled-components';

export const TargetItemContainer = styled.div`
`;
TargetItemContainer.displayName = 'TargetItemContainer';