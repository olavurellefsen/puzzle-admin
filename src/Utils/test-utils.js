import React from "react";
import { render } from "@testing-library/react";
import { ApolloProvider } from "react-apollo";
import MainContextStore from "../MainContextStore";
import client from "../heyBreydLocalizationClient";

const AllTheProviders = ({ children }) => {
  return (
    <ApolloProvider client={client}>
      <MainContextStore>{children}</MainContextStore>
    </ApolloProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
