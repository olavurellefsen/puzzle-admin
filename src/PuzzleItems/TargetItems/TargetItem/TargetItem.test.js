import React from "react";
import { render, cleanup } from 'test-utils';
import TargetItem from "./TargetItem";
afterEach(cleanup)

it("should render correctly", () => {
  const { container } = render(<TargetItem />);
  expect(container).toMatchSnapshot();
});

it("should show the TargetItemContainer element", () => {
  const { queryByTestId } = render(<TargetItem />);
  expect(queryByTestId("TargetItemContainer")).toBeTruthy();
});