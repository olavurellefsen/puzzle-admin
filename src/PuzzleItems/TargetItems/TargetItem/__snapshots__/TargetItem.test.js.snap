// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div
    class="sc-bdVaJa fePDys"
    data-testid="TargetItemContainer"
  />
</div>
`;
